lib/git_worktree.rb

```diff
@@ -0,0 +1,336 @@
+  end
+
+  def file_attributes(fname)
+    walker = Rugged::Walker.new(@repo)
+    walker.sorting(Rugged::SORT_DATE)
+    walker.push(@repo.ref(HEAD_REF).target)
+    commit = walker.find { |c| c.diff(:paths => [fname]).size > 0 }
+    return {} unless commit
+    {:updated_on => commit.time.gmtime, :updated_by => commit.author[:name]}
+  end
+
+  def merge(commit, rebase = false)
+    master_branch = @repo.ref(HEAD_REF)
+    merge_index = master_branch ? @repo.merge_commits(master_branch.target, commit) : nil
+    if merge_index && merge_index.conflicts?
+      result = differences_with_master(commit)
+      raise MiqException::MiqGitConflicts, result
+    end
+    commit = rebase(commit, merge_index, master_branch ? master_branch.target : nil) if rebase
+    @repo.reset(commit, :soft)
+  end
+
+  def rebase(commit, merge_index, parent)
+    commit_obj = commit if commit.class == Rugged::Commit
+    commit_obj ||= @repo.lookup(commit)
+    Rugged::Commit.create(@repo, :author    => commit_obj.author,
+                                 :committer => commit_obj.author,
+                                 :message   => commit_obj.message,
+                                 :parents   => parent ? [parent] : [],
+                                 :tree      => merge_index.write_tree(@repo))
+  end
+
+  def merge_and_push(commit)
+    rebase = false
+    push_lock do
+      @saved_cid = @repo.ref(HEAD_REF).target.oid
+      merge(commit, rebase)
+      rebase = true
+      @repo.push(@remote_name, [HEAD_REF], :credentials => @cred)
+    end
+  end
+
+  def fetch
+    options = {:credentials => @cred}
+    ssl_no_verify_options(options) do
+      @repo.fetch(@remote_name, options)
+    end
+  end
+
+  def pull
+    lock { fetch_and_merge }
+  end
+
+  def fetch_and_merge
+    fetch
+    commit = @repo.ref("refs/remotes/#{@remote_name}/master").target
+    merge(commit)
+  end
+
+  def file_list(commit_sha = nil)
+    tree = lookup_commit_tree(commit_sha || @commit_sha)
+    return [] unless tree
+    tree.walk(:preorder).collect { |root, entry| "#{root}#{entry[:name]}" }
+  end
+
+  def find_entry(path, commit_sha = nil)
+    get_tree_entry(path, commit_sha)
+  end
+
+  def mv_file_with_new_contents(old_file, new_path, new_data, commit_sha = nil, default_entry_keys = {})
+    add(new_path, new_data, commit_sha, default_entry_keys)
+    remove(old_file, commit_sha)
+  end
+
+  def mv_file(old_file, new_file, commit_sha = nil)
+    entry = current_index[old_file]
+    return unless entry
+    entry[:path] = new_file
+    current_index(commit_sha).add(entry)
+    remove(old_file, commit_sha)
+  end
+
+  def mv_dir(old_dir, new_dir, commit_sha = nil)
+    old_dir = fix_path_mv(old_dir)
+    new_dir = fix_path_mv(new_dir)
+    updates = current_index(commit_sha).entries.select { |entry| entry[:path].start_with?(old_dir) }
+    updates.each do |entry|
+      entry[:path] = entry[:path].gsub(old_dir, new_dir)
+      current_index(commit_sha).add(entry)
```

Comment:
Should we be ensuring you're not moving over the top of an existing file / directory? I guess a direct name conflict would cause rugged to complain... but accidentally merging two directories just because none of their contents conflict, sounds undesirable.